envi/__init__.py

```diff
@@ -904,13 +906,21 @@ class CallingConvention(object):
     arg_def = []
     retval_def = (CC_STACK, 0)
     retaddr_def = (CC_STACK, 0)
+    retvoid = False
```

Comment:
whether a function returns void/NULL doesn't have anything to do with the calling convention, this is a per-function value.  
if it's beneficial to store this in a CallingConvention instance, please make this a per-object setting.

whether a function returns VOID should be identified during analysis (eg. analysis/i386/calling.py or analysis/amd64/emulation.py) and stored using setFunctionMeta.  
this can also be identified using the impapi subsystem, but even then (iinm)  it should ultimately be stored in funcmeta as an attribute of that function.

Reply:
I don't think the impapi subsystem stores any function metadata, at least not for WinAPI.  It only stores the API (in analysis/__init__.py by calling addImpApi).  That's why I use getCallApi() in emulator.py to figure out if there is a return value.  I can probably change impapi to store using setFunctionMeta; is that your preference?  (because the main design principle is to store everything in funcmeta?)

Reply:
the calling convention doesn't store anything.  the funcmeta stores the details about what calling convention it is, as well as the api.  
impapi doesn't store anything.  it provides data about the functions.

setFunctionApi() is used to store the function api metadata, including rettype (aka. 'void')
setFunctionApi() stores this information stored in funcmeta as 'api', accessible using getFunctionApi().

getImpApiRetType() also returns the return-type.  

this information all needs to come together in a cohesive fashion, which is why vivisect.analysis.generic.impapi does just that.  it checks functions that have an import api defined, and creates an api definition using setFunctionApi()

does that context alter the conversation any?